puzzle-utils/encoders/caesar.py

```diff
@@ -14,22 +14,54 @@ def shift_letter(letter, shift_count):
     return chr(char_code + CHARACTER_OFFSET)
 
 def caesar(word, shift_count):
+    """
```

Comment:
Would you mind following the commenting styles of [restructured text](https://thomas-cokelaer.info/tutorials/sphinx/docstring_python.html)? It's one of the most commonly used ones out there and I'd like standardize on something.

Reply:
```
        **Advantages**:
         - Uses sphinx markups, which will certainly be improved in future
           version
         - Nice HTML output with the See Also, Note, Warnings directives


        **Drawbacks**:
         - Just looking at the docstring, the parameter, type and  return
           sections do not appear nicely
```
Do you think we'll be able to use the advantages? If we're just looking at this with help() or something, I think this doesn't look very good. You know what they say, the nice thing about standards is that there are so many to choose from! What about [this one](https://github.com/google/styleguide/blob/gh-pages/pyguide.md)?

Reply:
Based on these threads

https://stackoverflow.com/questions/3898572/what-is-the-standard-python-docstring-format
https://stackoverflow.com/questions/9195455/how-to-document-a-method-with-parameters

reST seems to be the defacto standard. It is even mentioned in [PEP 287](https://www.python.org/dev/peps/pep-0287). One of the important features I want for documentations is type annotations. Although the Google's style guide does support types (though only if we enhance it with this spec http://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html), there isn't a linter out there for Google style docs AFAIK. But there is one for reST style. See
https://gitlab.com/pycqa/flake8/issues/337 and
https://github.com/peterjc/flake8-rst-docstrings

Based on these findings I'm leaning towards reST atm. What do you think?

Reply:
I think that reST is ugly to look at, the fact that Google has and uses a standard means that said standard is pretty well established, your second stackoverflow link only lists reST as an option in a secondary answer so I'm not sure how it adds to your case, the main advantage of reST that we'd even use is to print nice things with Sphinx ([which supports Google style docstrings](http://www.sphinx-doc.org/en/master/usage/extensions/napoleon.html?highlight=google)), and the only reason to use a linter for this particular project would be because you want to do it for fun (which I can accept).

If you want a docstring linter for fun and you'd like to use reStructuredText for that, I'm happy to update this :). If you really just want to use it just to get into the habit for future projects, that's fine too and I don't mind going with that. There's lots of other reasons you could come up with for this. I just think that if we're looking at this from a purely practical point of view for this project, there isn't a good reason to use reStructuredText. Its verbosity is overkill and it is not as easy to read the plaintext as the alternatives.

Reply:
*Clarification:* A linter for code is useful because it can catch legit errors, one for docstrings is less so and only affects legibility (which we can easily monitor ourselves)

Reply:
Personally, the values I see in linters is that it's automatically enforces a style so that we can have a consistent look and feel for code base. In code reviews, any whitespace misalignment or mis-capitalized camelCase really sticks out like a sore thumb, but no reviewers want to spend the time or effort to point it out. That why I want linters to do that for me. Documentation is very much part of this as well.

I am aware that Sphinx supports Google's doc style. Though further investigation reveals that the reST linter https://github.com/peterjc/flake8-rst-docstrings is quite a weak linter for function docstrings. E.g. I don't believe it can catch errors like the parameters documentation not matching the function signature. So in that case I'm fine with the Google style. Although I'd like to point out

1. We *need* type information in function docstrings. Please specify them in the parameter documentation. Another option is to use [python's newly introduced typing annotations](https://docs.python.org/3/library/typing.html). We'd have to pick one here, and as I'm writing this I'm starting to lean towards the latter due to the abundance of checkers and enforcers around it. What do you think?
1. The code base already contains documentation using reST style. Adopting a Google style would be mean a migration effort in the future.
1. You might be aware of this but the code in the current PR is not adhering to the Google style.

Reply:
I agree that linters are useful. In particular, for Python, the Google style guide states about pylint that `It finds problems that are typically caught by a compiler for less dynamic languages like C and C++.` By itself this is a great asset. I just mean that I think having that same thing for docstrings is more of a nice-to-have than a must-have.

As for the points you've mentioned:
1. I totally agree. It is incredibly useful to know what a function's arguments are used for and their type since that is so hard to figure out in Python. I think the typing annotations you pointed out is an excellent suggestion and we should use it in addition to verbose documentation.
2. I don't mind updating the rest of the world since the rest of the world is currently fairly small I could do that in this same PR. As I mentioned before, I would also use reST if you really want to. After all, you _do_ own this repo 😃 . For me, this discussion was more about understanding the options, their benefits and downsides, and talking about why we'd pick one vs the other(s).
3. Yes, I'm aware of this 😜. In fact, I didn't know any of these standards even existed before this discussion. It has been very educational for me!

Alright, executive decision time. What'll it be?